### ISSUE TITLE
delay on notify

### ISSUE THREAD

**Issue by jgallen23** [jwilder/docker-gen#43]
my server takes a little bit of time to start up, so I'd like to have a delay before notifying nginx and adding to the proxy.  This sort of ties into #42, as I'd only want a single event if multiple were fired within the delay interval

**Comment by RoboMod** [comment]
+1

**Comment by jgallen23** [comment]
the other thing I was thinking, is it would be nice to have docker-gen do a simple http check against the container and not notify nginx until it returns a 200.  if it returns a 5xx, then don't add to nginx.  Thoughts?

**Comment by fazy** [comment]
It seems to me that docker-gen already does two things (simplifying a bit):

1. Watch Docker containers for events (start, stop)
2. Generate config file (and call any kind of reload method as needed)

The issue is when step 2 happens too soon after step 1. We need an intermediate step. This could be, for example:

1. Wait a certain length of time
2. Check a socket (TCP/UDP port *n*)
3. Check a service (HTTP 200 OK)
4. Custom check

You could argue that docker-gen should not take on this responsibility, and if so then the last one, custom check, might be the best option.

I'd be happy for this feature to be implemented like this:

Custom check = a specified command executable. The command is expected to wait for whatever condition is needed to satisfy that the newly created Docker container is fully running. After some time, it should return exit code 0, and then docker-gen can get on with generating the config file.

A non-zero exit code would cause docker-gen to not load the config.

I guess docker-gen would need to call the custom command with some details about the docker container, IP address for sure, maybe port.

Is this likely to work?

**Comment by jwilder** [comment]
I think #67 is general enough that you could implement a delay, HTTP poll, socket test, config check, etc.

**Comment by fazy** [comment]
Yes, that would probably work in most cases.

However, I'm not sure what happens if you write the config file first, as suggested in #67. If two containers start, and one takes a second to become "ready" and the other takes 5 seconds, then the config update/reload would, for a short time, activate the config for the slower container before it's ready.

**Comment by jwilder** [comment]
Fixed via #156
